### PR TITLE
[gawb-3964] automation test for sam policies and google acls

### PR DESCRIPTION
**Pull request by marctalbott**
closed and reopened because jenkins automation tests were not starting

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala [hunk 1]
```diff
@@ -351,5 +353,116 @@ class RawlsApiSpec extends TestKit(ActorSystem("MySpec")) with FreeSpecLike with
...
+      val workspacePolicyNames = Set("can-compute", "project-owner", "writer", "reader", "share-writer", "can-catalog", "owner", "share-reader")
+      // bucket and object access levels for sam policies as described in comments in insertBucket function in HttpGoogleServicesDAO
+      val policyToBucketAccessLevel = Map("project-owner" -> "WRITER", "owner" -> "WRITER", "writer" -> "WRITER", "reader" -> "READER")
+      val policyToObjectAccessLevel = Map("project-owner" -> "READER", "owner" -> "READER", "writer" -> "READER", "reader" -> "READER")
+      val serviceAccountEmailDomain = "developer.gserviceaccount.com"
+
+      withCleanBillingProject(owner) { projectName =>
+        withWorkspace(projectName, s"unconstrained-workspace") { workspaceName =>
+          val workspaces = Rawls.workspaces.list()
+          println(workspaces)
```
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala]: leftover `println` or do you need to log this?

Comment (marctalbott) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala]: leftover! thanks

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala [hunk 2]
```diff
@@ -351,5 +353,116 @@ class RawlsApiSpec extends TestKit(ActorSystem("MySpec")) with FreeSpecLike with
...
+      implicit val token: AuthToken = ownerAuthToken
+      val workspacePolicyNames = Set("can-compute", "project-owner", "writer", "reader", "share-writer", "can-catalog", "owner", "share-reader")
+      // bucket and object access levels for sam policies as described in comments in insertBucket function in HttpGoogleServicesDAO
+      val policyToBucketAccessLevel = Map("project-owner" -> "WRITER", "owner" -> "WRITER", "writer" -> "WRITER", "reader" -> "READER")
+      val policyToObjectAccessLevel = Map("project-owner" -> "READER", "owner" -> "READER", "writer" -> "READER", "reader" -> "READER")
+      val serviceAccountEmailDomain = "developer.gserviceaccount.com"
+
+      withCleanBillingProject(owner) { projectName =>
+        withWorkspace(projectName, s"unconstrained-workspace") { workspaceName =>
+          val workspaces = Rawls.workspaces.list()
```
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala]: Just a thought, at some point we should add a getWorkspaceDetails(namespace, name) to [Rawls](https://github.com/broadinstitute/workbench-libs/blob/develop/serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/Rawls.scala#L33), for when we want just one particular workspace, but doesn't have to be for this ticket

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala [hunk 3]
```diff
@@ -351,5 +353,116 @@ class RawlsApiSpec extends TestKit(ActorSystem("MySpec")) with FreeSpecLike with
...
+      withCleanBillingProject(owner) { projectName =>
+        withGroup("authDomain", List(owner.email)) { authDomain =>
+          withWorkspace(projectName, s"constrained-workspace", Set(authDomain)) { workspaceName =>
+            val workspaces = Rawls.workspaces.list()
+            val workspaceId = workspaces.parseJson.asInstanceOf[JsArray].elements.flatMap { workspaceListEntry =>
+              workspaceListEntry.asJsObject.getFields("workspace").collect {
+                case workspace if (workspace.asJsObject.getFields("name").exists(name => name.convertTo[String].equals(workspaceName)) &&
+                  workspace.asJsObject.getFields("namespace").exists(namespace => namespace.convertTo[String].equals(projectName))) => workspace.asJsObject.getFields("workspaceId").head
+              }
+            }.head.convertTo[String]
```
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala]: There are some repeated blocks of code in these two tests (like parsing out the workspace Ids) that you could pull out into a separate function

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala [hunk 4]
```diff
@@ -351,5 +353,108 @@ class RawlsApiSpec extends TestKit(ActorSystem("MySpec")) with FreeSpecLike with
...
+    }
+
+    "should have correct policies in Sam and ACLs in Google when a constrained workspace is created" in {
+      implicit val patienceConfig: PatienceConfig = PatienceConfig(timeout = 20 seconds)
+      implicit val token: AuthToken = ownerAuthToken
+      val workspacePolicyNames = Set("can-compute", "project-owner", "writer", "reader", "share-writer", "can-catalog", "owner", "share-reader")
+      // bucket and object access levels for sam policies as described in comments in insertBucket function in HttpGoogleServicesDAO
+      val policyToBucketAccessLevel = Map("project-owner" -> "WRITER", "owner" -> "WRITER", "writer" -> "WRITER", "reader" -> "READER")
+      val policyToObjectAccessLevel = Map("project-owner" -> "READER", "owner" -> "READER", "writer" -> "READER", "reader" -> "READER")
+      val serviceAccountEmailDomain = "developer.gserviceaccount.com"
```
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsApiSpec.scala]: Repeated blocks (workspacePolicyNames, policyToBucketAccessLevel, policyToObjectAccessLevel, serviceAccountEmailDomain) here can also be set once outside the tests and used in both places.

**Review by ansingh7115 — APPROVED** [review]

**Pull request merged**